prov/efa/src/efa_mr.c

```diff
@@ -498,7 +501,7 @@ static int efa_mr_dereg_impl(struct efa_mr *efa_mr)
 		efa_mr->shm_mr = NULL;
 	}
 
-	if (efa_mr->peer.iface == FI_HMEM_CUDA && cuda_is_gdrcopy_enabled()) {
+	if (efa_mr->peer.iface == FI_HMEM_CUDA && !ofi_cudaIsDeviceId(efa_mr->peer.device.cuda)) {
```

Comment:
This sort of check looks like an incorrect use of the APIs.  Why would the device value be invalid?  If it isn't being set, how do you know what's there isn't accidentally valid?

Reply:
EFA will set `device.cuda` as gdr handle when gdrcopy is available. We used to use this to check whether `device` is a handle or a device id. But your point is totally right that this logic is problematic. We could introduce explicit check whether gdrcopy is being used so this `ofi_cudaIsDeviceId` call can be avoided.